src/common/misc.c

```diff
@@ -223,6 +223,28 @@ daos_sgls_buf_size(d_sg_list_t *sgls, int nr)
 	return size;
 }
 
+int
+daos_sgl_buf_extend(d_sg_list_t *sgl, int idx, size_t new_size)
+{
+	char	*new_buf;
+
+	if (sgl == NULL || sgl->sg_iovs == NULL)
+		return 0;
+
+	D_ASSERT(sgl->sg_nr > idx);
+	if (sgl->sg_iovs[idx].iov_buf_len >= new_size)
+		return 0;
+
+	D_REALLOC(new_buf, sgl->sg_iovs[idx].iov_buf, new_size);
```

Comment:
does the old buf get freed in this case?  If realloc fails it is untouched...but does it get freed in the error path?

Reply:
if extend fails, the old buffer will still there. and yes, it will be freed in the error path.